src/librbd/cache/ObjectCacherObjectDispatch.h

```diff
@@ -49,6 +49,15 @@ class ObjectCacherObjectDispatch : public io::ObjectDispatchInterface {
       io::DispatchResult* dispatch_result, Context** on_finish,
       Context* on_dispatched) override;
 
+  bool read_extents_with_version(
```

Comment:
I would be more interested to see the existing `read` interface be updated to include `Extents`. Off the top of my head I don't think you would need to pass a bufferlist and ExtentMap for each extent since the bufferlist+ExtentMap are already a sparse data structure. 

Using the RADOS object versioning is an interesting approach to handle multiple librbd clients writing to the same image concurrently using unaligned extents. If you were just interested in avoiding the same librbd client causing an overwrite during a RMW cycle, it would be more efficient to use the `BlockGuard` to pause overlapping IO temporarily. 

I was originally thinking you could use a (tweaked) compare-and-write to avoid multi-client overwrites but assuming this works as it should, this would be more efficient and easier to understand.

Reply:
But if a non-sparse read is used, we still need multiple bufferlists, right?

BTW for crypto we will need to assemble the sparse bufferlists before we call the decrypt function.